happi/audit.py

```diff
@@ -0,0 +1,424 @@
+"""
+This module defines the ``happi audit`` command line utility
+
+If the --file option is not provided, then it will take the path from
+Happi configuration file using the Client.find_config method,
+otherwise it will use the path specified after --file
+"""
+from abc import ABCMeta, abstractmethod
+from configparser import ConfigParser
+import logging
+import os
+import sys
+import subprocess
+import re
+from happi.client import Client
+from happi.loader import import_class, fill_template
+from happi.containers import registry
+from enum import Enum
+
+logger = logging.getLogger(__name__)
+
+
+class ReportCode(Enum):
+    """
+    Class to define report codes
+    """
+    SUCCESS = 0
+    INVALID = 1
+    MISSING = 2
+    EXTRAS = 3
+    NO_CODE = 9
+
+
+class Command(object, metaclass=ABCMeta):
+    """
+    Bluprint for the command class
+    This will be useful if all the commands are inheriting from it
+    """
+
+    def __init__(self, name, summary):
+        self.name = name
+        self.summary = summary
+
+    @abstractmethod
+    def add_args(self, parser):
+        """
+        Adds arguments to a parser
+
+        Parameters
+        -----------
+        parser: ArgumentParser
+            Parser to add arguments to
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def run(self, args):
+        """
+        Parses the arguments given to the command.
+        And hanles the logic for this command
+
+        Parameters
+        -----------
+            args: Namespace
+        """
+        raise NotImplementedError
+
+
+class Audit(Command):
+    """
+    Audits the database for valid entries.
+    """
+    def __init__(self):
+        self.name = 'audit'
+        self.help = "Inspect a database's entries"
+        self._all_devices = set()
+        self._all_items = []
+        # 9 - NO_CODE
+        self.report_code = ReportCode(9)
+
+    def add_args(self, cmd_parser):
+        cmd_parser.add_argument(
+            "--file", help='Path to the json file (database) to be audited'
+        )
+        cmd_parser.add_argument(
+            "--extras", action='store_true', default=False,
+            help='Check specifically for extra attributes'
+        )
+
+    def run(self, args):
+        """
+        Validate the database passed in with --file option
+        """
+        if args.file is not None:
```

Comment:
While it fits in with your `Command` class and it makes sense in this context, consider the following:

I want to use a Python script to validate 30 different databases without resorting to shelling out to `happi audit ...`.
How do I do that with this code? Instantiate this `Audit()` class, fake out the `parse_args()`, and then call `run()`?

The functionality of the class should be entirely broken out into separate, independently usable functions, not relying on the state held in the class. Additionally, the CLI interface to these functions should be relatively small.

Please let me know if you need further details here.

Reply:
Ohh I did not consider that...... 
I have implemented some changes related to this in 7cf8a0b, but I need to revisit it..... (especially if I implement your suggestion above https://github.com/slaclab/pytmc/blob/master/pytmc/bin/pytmc.py#L27  )
`run` and `add_args` should be the only ones dealing with `cli` now.... I will get back to you with this one. 

Thanks a lot for pointing this out..